Support plugins extending Webpack build pipeline; Build translations from openproject-translations

myabc opened the pull request:
This PR concerns itself with two related, but separate themes:
### Support plugins extending Webpack build pipeline

https://community.openproject.org/work_packages/5697
https://community.openproject.org/work_packages/17241
#### Notes for review:
- Webpack will not bundle plugins (and fail silently) if a `Gemfile.lock` is inconsistent (for example, if there are conflict makers). I'm not how/if we should make this more obvious.
- Shelling out from node to Ruby is currently done **synchronously**, which is not really _the node way_, but I wasn't really sure how to handle callbacks in the webpack config.
### Build translations from openproject-translations, if available

https://community.openproject.org/work_packages/14679
#### Notes for review:
- ~~You can specify an `OPENPROJECT_TRANSLATIONS_ROOT` if you want to
  test this PR without adding openproject-translations as a dependency.~~
  
  _Removed in 4ea7cb0. Use a bundle override instead (e.g. `bundle config local.openproject-translations ~/git-repos/openproject-translations`_

---

See also 
- finnlabs/openproject-backlogs#114
- finnlabs/openproject-costs#115

myabc committed:
Do not specify Webpack loaders in requires

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Specify resolve.root, DRY paths in Webpack config

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Teach Webpack to bundle translations plugin locales

Bug # 14679

Signed-off-by: Alex Coles <alex@alexbcoles.com>

TeatroIO commented:
I've prepared a stage. <a href='http://feature-webpack-i18n-support.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

myabc committed:
Expose I18n as global with Webpack expose-loader

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@ulferts @tessi How do we support translations for plugins that are neither German or English?

|  | core app | plugins |
| --- | --- | --- |
| bundled translations |  |  |
| en | X | X |
| de | X | X |
| openproject-translations |  |  |
| it | X | ? |
| es | X | ? |
| cn | X | ? |
| etc. | X | ? |

ulferts commented:
@myabc don't know that, sorry.

myabc committed:
Add shelljs for synchronous command exec

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add helpers to make node translations-plugin aware

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Remove i18n-js (as gem dependency)

Translations are now bundled by Webpack.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
Hound will continue to bark until #2184 is merged ... hint, hint... :poodle:

tessi commented:
@myabc @ulferts I had a little chat with @kgalli about including plugin translations yesterday. He suggested to split the translation files from the plugin-code (and host the translations on some separate opf-related service). This would enable us to have many crowdin projects (and other translation services would probably work too), so that we can have one crowdin-project per plugin.

The translations-plugin could then fetch the translations of the current OpenProject version from that opf-service (which in turn asks crowdin) for the core and all plugins. In a second step, those translations could be combined into one `<locale>.yml` file.

You can probably ask @kgalli for details.

myabc committed:
Flesh out helpers to support all OP(Rails) plugins

This currently use a very native approach to determines if a gem is a
OpenProject plugin - simply matching 'openproject-'.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add lodash (as npm dependency)

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add Rails plugin dirs to Webpack modulesDirectories

This should allow Webpack to bundle assets from Rails plugins/gems,
as if they were npm modules or Bower components, but without us
having to add an entry to core's `package.json`/`bower.json`.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add I18n function for plugins to add translations

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Create Webpack entry points for each plugin

User Story # #5697

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Filter out OP(Rails) plugins without package.json

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update paths in Protractor HTML file, gulp tasks

Changed in previous commit 3d34ff2e.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Stick to single quotes in webpack.config.js

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Allow a few shelljs globals in rails-plugin.conf.js

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Move Angular app under frontend/app

This mass rename has been brought forward to work-around an apparent
naming collision when running Karma tests.

    ERROR [karma]: [Error: Path is a file /_js/app/assets/javascripts/angular]
    Error: Path is a file /_js/app/assets/javascripts/angular

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Merge branch 'dev' into feature/webpack-i18n-support

myabc committed:
Move auto-complete-helper.js into place

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Teach Webpack to build Bundler git dependencies

Plugins that are Bundler git dependencies have a SHA fragment appended
to the directory where they are cloned, e.g.
`openproject-backlogs-e8338514e9d9`.

We create an alias to help Webpack handle this.

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
[wip] Add executable to list all OP(Rails) plugins

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Update helpers to use plugin list executable

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add Copyright headers to rails-plugins.conf.js

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Add Copyright headers to plugin_info

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
I've added notes to the description.

There are some additional, related tasks that I'll leave for other Pull Requests:
- bundling HTML templates along with plugin JS assets.
- moving Karma and Protractor tests under `frontend`

myabc committed:
Add initial docs for developing, using PLUGINS

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
Additional, related tasks for translations that I'll leave for other Pull Requests:
- use  CrowdIn for German, English core translations.
- setting up CrowdIn projects for individual plugins.

myabc committed:
Add note on assets, translations to PLUGINS

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

ulferts commented on bin/plugin_info:
```diff
@@ -26,25 +28,33 @@
+  end
+end
+
+puts JSON.generate(plugin_name_paths)
```
@myabc I don't think this is intended.

myabc replied:
@ulferts it is intended. The script should write to stdout. I guess `STDOUT.puts` might make the intention clearer.

ulferts replied:
@myabc what is intended to happen with that output? Why is it required to write to stdout? Why can't the loader of that file handle all of this internally?

myabc replied:
@ulferts `config/rails-plugins.conf.js` currently shells out to this script. If you can think a way of improving this, let me know.

ulferts replied:
I see... you have to get the output from ruby to js. No clean solution for that comes to mind. We could wrap it all in function that one can call with parameters. If no param is provided, nothing happens. The provided parameter specifies the means for transporting the information. But that would basically still be the same puts statement. Could be an ENV variable just the same but this does not make it better.

ulferts commented:
looks good to me and works on my machine.

0xF013 commented:
Looks ok to me

myabc committed:
Fix small typo in PLUGINS

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Treat translations plugin as any other OP plugin

Replace special handling with simple entry point in
openproject-translations plugin.
See https://github.com/opf/openproject-translations/commit/4fe08c6e646a598e2e055ba7c7b4b938a3a9516b

Bug # 14679

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Fix JSHint unused warnings in rails-plugin.conf.js

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc committed:
Rename PLUGINS->DEVELOP_PLUGINS to stress audience

Add note that pure frontend plugins are a WIP.

[ci skip]

Signed-off-by: Alex Coles <alex@alexbcoles.com>

myabc commented:
@ulferts issues you mentioned should be dealt with.

myabc committed:
Merge branch 'dev' into feature/webpack-i18n-support

ulferts commented:
Merging because I know that the failing cukes have nothing to do with this PR.
